src/java/org/apache/cassandra/gms/Gossiper.java

```diff
@@ -158,38 +158,44 @@
 
     private volatile long lastProcessedMessageAt = System.currentTimeMillis();
 
-    //This property and anything that checks it should be removed in 5.0
-    private boolean haveMajorVersion3Nodes = true;
+    // This property and anything that checks it should be removed in 5.0
+    /**
+     * This property is initially set to null version which means that we have no information about the other nodes.
+     * Once we have information about the cluster, it is set to the earliest version encountered in the cluster.
+     * Once all nodes are on at least this node version, it becomes null, which means that we are not upgrading from
+     * the previous version (major, minor).
+     */
+    private CassandraVersion upgradeFromVersion = SystemKeyspace.NULL_VERSION;
 
-    final Supplier<ExpiringMemoizingSupplier.ReturnValue<Boolean>> haveMajorVersion3NodesSupplier = () ->
+    final Supplier<ExpiringMemoizingSupplier.ReturnValue<CassandraVersion>> upgradeFromVersionSupplier = () ->
```

Comment:
Somehow the comment lost... 

One concern about the new supplier is that it is more expensive, since the prior version has a shortcut to exit early, but the new one has to iterate through all peers to find the `minVersion`. Should it have a higher chance to encounter the case that `getReleaseVersion` returns `null` and delay calculating the `minVersion`?

Reply:
the comment was moved a bit, but not lost

do you think we can safely assume there are maximum two different versions in the cluster? If so, we can break the loop once we find the first version which is lower than this node version

Reply:
The cluster wide version check is a special treatment for specific versions, i.e. 3.4 and `v < 4`. What if we keep the checker as-is and always using the `3.4-` behavior during upgrade? 
It could be more expensive (by how much? probably not a lot) during that time period, but the gain is the simplification. 

In practice, I think there won't be a case that the upgrading cluster have nodes in `3.4-`, `3.11` and `4.0`. The upgrading path is most likely either 3.0 -> 4.0, or 3.x -> 4.0. With such assumption, if we want to lower the data size being queried, we can define a temporarily system property in the 4.0 version to select the behavior. e.g. `cassandra.upgrading_from_3_0_plus`, according to the value use the corresponding behavior. The property, the same as `Gossiper#upgradeFromVersion`, should be removed in 5.0. If we are adding the property, we need to document it in the `NEWS.txt` too in order to inform the instructor when upgrading. 
(Maybe just using the `3.4-` behavior is the best option, as it add complexity in operations)

Reply:
I was thinking about that but came into conclusion that unfortunately we could not do that. Look, while 4.0 notices the upgrade from 3.11 and it can act in whichever way we decide, 3.11 will work as usual and use 3.11 `ColumnFilter`. That's why we cannot just simply make 4.0 work in 3.0 compatibility mode when upgrading from any 3.0 or 3.11.

Reply:
Fair. It is messy and hopefully we can do better in the later releases.